### PR TITLE
kv: deflake TestClientNotReady

### LINKED ISSUE #5187: stress: failed test in cockroach/kv/kv.test: TestClientNotReady
Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/6a7f101fdcc8adda1e84d85b5546396964b622fa

Stress build found a failed test:

```
=== RUN   TestClientNotReady
I160312 05:14:07.366731 /go/src/google.golang.org/grpc/server.go:454  grpc: Server.processUnaryRPC failed to write status stream error: code = 1 desc = "context canceled"
--- FAIL: TestClientNotReady (0.02s)
	send_test.go:286: kv/send_test.go:265: unexpected error when client was closed: failed to send RPC: too many errors encountered (1 of 1 total): rpc error: code = 2 desc = "grpc: the client connection is closing\n```

Run Details:

```
0 runs so far, 0 failures, over 5s
0 runs so far, 0 failures, over 10s
1 runs so far, 0 failures, over 15s
8 runs so far, 0 failures, over 20s
8 runs so far, 0 failures, over 25s
12 runs so far, 0 failures, over 30s
16 runs so far, 0 failures, over 35s
16 runs so far, 0 failures, over 40s
20 runs so far, 0 failures, over 45s
23 runs so far, 0 failures, over 50s
24 runs so far, 0 failures, over 55s
26 runs so far, 0 failures, over 1m0s
32 runs so far, 0 failures, over 1m5s
33 runs so far, 0 failures, over 1m10s
35 runs so far, 0 failures, over 1m15s
37 runs so far, 0 failures, over 1m20s
41 runs so far, 0 failures, over 1m25s
42 runs so far, 0 failures, over 1m30s
46 runs so far, 0 failures, over 1m35s
49 runs so far, 0 failures, over 1m40s
51 runs so far, 0 failures, over 1m45s
54 runs so far, 0 failures, over 1m50s
57 runs so far, 0 failures, over 1m55s
60 runs so far, 0 failures, over 2m0s
63 runs so far, 0 failures, over 2m5s
65 runs so far, 0 failures, over 2m10s
67 runs so far, 0 failures, over 2m15s
70 runs so far, 0 failures, over 2m20s
73 runs so far, 0 failures, over 2m25s
75 runs so far, 0 failures, over 2m30s
78 runs so far, 0 failures, over 2m35s
82 runs so far, 0 failures, over 2m40s
84 runs so far, 0 failures, over 2m45s
87 runs so far, 0 failures, over 2m50s
90 runs so far, 0 failures, over 2m55s
92 runs so far, 0 failures, over 3m0s
96 runs so far, 0 failures, over 3m5s
98 runs so far, 0 failures, over 3m10s
100 runs so far, 0 failures, over 3m15s
105 runs so far, 0 failures, over 3m20s
107 runs so far, 0 failures, over 3m25s
107 runs so far, 0 failures, over 3m30s
112 runs so far, 0 failures, over 3m35s
115 runs so far, 0 failures, over 3m40s
115 runs so far, 0 failures, over 3m45s
120 runs so far, 0 failures, over 3m50s
121 runs so far, 0 failures, over 3m55s
124 runs so far, 0 failures, over 4m0s
128 runs so far, 0 failures, over 4m5s
129 runs so far, 0 failures, over 4m10s
132 runs so far, 0 failures, over 4m15s
136 runs so far, 0 failures, over 4m20s
137 runs so far, 0 failures, over 4m25s
139 runs so far, 0 failures, over 4m30s
143 runs so far, 0 failures, over 4m35s
145 runs so far, 0 failures, over 4m40s
147 runs so far, 0 failures, over 4m45s
151 runs so far, 0 failures, over 4m50s
153 runs so far, 0 failures, over 4m55s
156 runs so far, 0 failures, over 5m0s
158 runs so far, 0 failures, over 5m5s
161 runs so far, 0 failures, over 5m10s
165 runs so far, 0 failures, over 5m15s
166 runs so far, 0 failures, over 5m20s
168 runs so far, 0 failures, over 5m25s
173 runs so far, 0 failures, over 5m30s
175 runs so far, 0 failures, over 5m35s
177 runs so far, 0 failures, over 5m40s
181 runs so far, 0 failures, over 5m45s
183 runs so far, 0 failures, over 5m50s
186 runs so far, 0 failures, over 5m55s
189 runs so far, 0 failures, over 6m0s
191 runs so far, 0 failures, over 6m5s
193 runs so far, 0 failures, over 6m10s
197 runs so far, 0 failures, over 6m15s
199 runs so far, 0 failures, over 6m20s
201 runs so far, 0 failures, over 6m25s
204 runs so far, 0 failures, over 6m30s
208 runs so far, 0 failures, over 6m35s
209 runs so far, 0 failures, over 6m40s
214 runs so far, 0 failures, over 6m45s
215 runs so far, 0 failures, over 6m50s
217 runs so far, 0 failures, over 6m55s
221 runs so far, 0 failures, over 7m0s
223 runs completed, 1 failures, over 7m4s
FAIL
```
Please assign, take a look and update the issue accordingly.

### PR DESCRIPTION
**Pull request by petermattis**
This test is checking behavior of timeouts and closed connections when
the server isn't responding. Unfortunately, recent refactorings changed
it so that a real server was responding to requests making the test
racy.

Fixes #5187.

---

This change is [Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5198)

### CODE REVIEW

File: kv/send_test.go [hunk 1]
```diff
@@ -259,9 +265,15 @@ func TestClientNotReady(t *testing.T) {
 	opts.SendNextTimeout = 0
 	opts.Timeout = 0
 	c := make(chan error)
+	sent := make(chan struct{})
 	go func() {
+		// This is inherently racy. We want to close the connection after the batch
+		// RPC is sent, so we delay 100ms from now to close the sent channel.
+		time.AfterFunc(100*time.Millisecond, func() {
```
Comment (tschottdorf) [inline kv/send_test.go]: you couldn't accept a connection on the listener to get rid of the raciness?

**Comment by tamird** [comment]
Thanks for digging into this.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[kv/send_test.go, line 233 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198#-KCgnLWAAt87IBTts25O:-KCgnLWAAt87IBTts25P:1271735120) ([raw file](https://github.com/cockroachdb/cockroach/blob/c018d7ae6c154d93ea5687e806e17bfed4f5a9c5/kv/send_test.go#L233)):*
can `s` just be removed?

---

*[kv/send_test.go, line 272 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198#-KCgV5E-CzWxkrBazi2Y-r1-272:-KCgnV1Lg57vXIredWWI:5172523) ([raw file](https://github.com/cockroachdb/cockroach/blob/c018d7ae6c154d93ea5687e806e17bfed4f5a9c5/kv/send_test.go#L272)):*
+1

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198#-:-KCgnW_AbVFVINa1M33X:2101604060)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[kv/send_test.go, line 233 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198#-KCgnLWAAt87IBTts25O:-KCkpxfdwfO1XDKmhde8:-1619964508) ([raw file](https://github.com/cockroachdb/cockroach/blob/c018d7ae6c154d93ea5687e806e17bfed4f5a9c5/kv/send_test.go#L233)):*
Yes, yes it can.

---

*[kv/send_test.go, line 272 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198#-KCgV5E-CzWxkrBazi2Y-r1-272:-KCksIcn4ud1Eye3wK-p:2107932915) ([raw file](https://github.com/cockroachdb/cockroach/blob/c018d7ae6c154d93ea5687e806e17bfed4f5a9c5/kv/send_test.go#L272)):*
Good idea. This required making sure we had a fresh `rpc.Context` for the request to ensure a new connection was created when sending the batch.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[kv/send_test.go, line 279 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198#-KCkuAQnMwsOn0ARQ2eU:-KCkuAQnMwsOn0ARQ2eV:80870934) ([raw file](https://github.com/cockroachdb/cockroach/blob/09cb1677e7133744b37e46ce35afd0edb79ab269/kv/send_test.go#L279)):*
looking at this again, I think this should be `"failed as client connection was closed"` only - if we get the grpc error then we're really exercising some other code path than the one in `sendOne`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[kv/send_test.go, line 279 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198#-KCkuAQnMwsOn0ARQ2eU:-KCkvzssoo-Ppi1S-BIF:1879663881) ([raw file](https://github.com/cockroachdb/cockroach/blob/09cb1677e7133744b37e46ce35afd0edb79ab269/kv/send_test.go#L279)):*
Ok, ran stress for 5min without failures. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198)*

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198#-:-KCkwP3B3wSZ2IsWu7yY:1417088721)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[kv/send_test.go, line 274 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198#-KCkwWdJzsv70sB9EJ6e:-KCkwWdJzsv70sB9EJ6f:1145207219) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff0de1224e7fe908590039c6934985de4d9d265/kv/send_test.go#L274)):*
optional nit: i think an error check here wouldn't hurt

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[kv/send_test.go, line 274 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198#-KCkwWdJzsv70sB9EJ6e:-KCkxR8REMIYZrMq9UKT:1660675680) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ff0de1224e7fe908590039c6934985de4d9d265/kv/send_test.go#L274)):*
Sure. Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5198)*

**Pull request merged**